ksqldb-api-client http2MultiplexingLimit unrecoverable exception

**Describe the bug**
In `ClientImpl.java`  `HttpClientOptions` does not set the `http2MultiplexingLimit` resulting in it being -1.
This in turn, is used here in Vertx

```
public static VertxHttp2ConnectionHandler<Http2ClientConnection> createHttp2ConnectionHandler(
  HttpClientImpl client,
  Object queueMetric,
  ConnectionListener<HttpClientConnection> listener,
  ContextInternal context,
  Object socketMetric,
  BiConsumer<Http2ClientConnection, Long> c) {
  long http2MaxConcurrency = client.getOptions().getHttp2MultiplexingLimit() <= 0 ? Long.MAX_VALUE : client.getOptions().getHttp2MultiplexingLimit();
```
`http2MaxConcurrency`is then Long.MAX_VALUE.

Later, in `Pool.java` in Vertx, in our case concurrency is set to 100 on connection upgrade and holder.concurrency at `Long.MAX_VALUE`.
```
  private synchronized void setConcurrency(Holder holder, long concurrency) {
    if (concurrency < 0L) {
      throw new IllegalArgumentException("Cannot set a negative concurrency value");
    }
    if (holder.removed) {
      assert false : "Cannot recycle removed holder";
      return;
    }
    if (holder.concurrency < concurrency) {
      long diff = concurrency - holder.concurrency;
      if (holder.capacity == 0) {
        available.add(holder);
      }
      capacity += diff;
      holder.capacity += diff;
      holder.concurrency = concurrency;
      checkProgress();
    } else if (holder.concurrency > concurrency) {
      throw new UnsupportedOperationException("Not yet implemented");
    }
  }
```

This results in the error also seen in #6728
In our case, we can reproduce this only on remote connections, not on local ones (both with Docker).


**To Reproduce**
- Build an example client with https://docs.ksqldb.io/en/latest/developer-guide/ksqldb-clients/java-client/
- Set up for a remote connection with KSQL Server 0.15
- Create a `CompletableFuture<StreamedQueryResult> streamedQuery`and `get()`it

**Expected behavior**
No exception.

**Actual behaviour**
```
Exception in thread "main" java.util.concurrent.ExecutionException: java.lang.UnsupportedOperationException: Not yet implemented
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999)
	at my.ksqldb.app.ExampleApp.main(ExampleApp.java:30)
Caused by: java.lang.UnsupportedOperationException: Not yet implemented
```

**Additional context**
I guess this is rather a Vertx issue than a KSQL client issue, but it's a blocker when using tools that rely on the HTTP/2 streaming API client.

--- a/ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/ClientOptions.java
+++ b/ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/ClientOptions.java
@@ -25,6 +25,7 @@ public interface ClientOptions {
   String DEFAULT_HOST = "localhost";
   int DEFAULT_HOST_PORT = 8088;
   int DEFAULT_EXECUTE_QUERY_MAX_RESULT_ROWS = 10000;
+  int DEFAULT_HTTP2_MULTIPLEXING_LIMIT = -1;
 
   /**
    * Sets the host name of the ksqlDB server to connect to. Defaults to "localhost".
@@ -135,6 +136,14 @@ public interface ClientOptions {
   ClientOptions setExecuteQueryMaxResultRows(int maxRows);
 
   /**
+   * Sets the maximum number of requests per HTTP/2 connection. Defaults to -1.
+   *
+   * @param http2MultiplexingLimit number of requests
+   * @return a reference to this
+   */
+  ClientOptions setHttp2MultiplexingLimit(int http2MultiplexingLimit);
+
+  /**
    * Returns the host name of the ksqlDB server to connect to.
    *
    * @return host name
@@ -238,6 +247,13 @@ public interface ClientOptions {
    * @return number of rows
    */
   int getExecuteQueryMaxResultRows();
+
+  /**
+   * Returns the maximum number of requests per HTTP/2 connection.
+   *
+   * @return number of requests
+   */
+  int getHttp2MultiplexingLimit();
 
   /**
    * Creates a copy of these {@code ClientOptions}.

--- a/ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/ClientImpl.java
+++ b/ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/ClientImpl.java
@@ -621,7 +621,8 @@ public class ClientImpl implements Client {
         .setHttp2ClearTextUpgrade(false)
         .setVerifyHost(clientOptions.isVerifyHost())
         .setDefaultHost(clientOptions.getHost())
-        .setDefaultPort(clientOptions.getPort());
+        .setDefaultPort(clientOptions.getPort())
+        .setHttp2MultiplexingLimit(clientOptions.getHttp2MultiplexingLimit());
     if (clientOptions.isUseTls() && !clientOptions.getTrustStore().isEmpty()) {
       final JksOptions jksOptions = VertxSslOptionsFactory.getJksTrustStoreOptions(
           clientOptions.getTrustStore(),

--- a/ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/ClientOptionsImpl.java
+++ b/ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/ClientOptionsImpl.java
@@ -234,7 +234,9 @@ public class ClientOptionsImpl implements ClientOptions {
   }
 
   @Override
-  public int getHttp2MultiplexingLimit() { return http2MultiplexingLimit; }
+  public int getHttp2MultiplexingLimit() {
+    return http2MultiplexingLimit;
+  }
 
   @Override
   public ClientOptions copy() {

--- a/ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/ClientOptionsImpl.java
+++ b/ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/ClientOptionsImpl.java
@@ -35,6 +35,7 @@ public class ClientOptionsImpl implements ClientOptions {
   private String basicAuthUsername;
   private String basicAuthPassword;
   private int executeQueryMaxResultRows = ClientOptions.DEFAULT_EXECUTE_QUERY_MAX_RESULT_ROWS;
+  private int http2MultiplexingLimit = ClientOptions.DEFAULT_HTTP2_MULTIPLEXING_LIMIT;
 
   /**
    * {@code ClientOptions} should be instantiated via {@link ClientOptions#create}, NOT via this
@@ -52,7 +53,7 @@ public class ClientOptionsImpl implements ClientOptions {
       final String trustStorePath, final String trustStorePassword,
       final String keyStorePath, final String keyStorePassword, final String keyPassword,
       final String keyAlias, final String basicAuthUsername, final String basicAuthPassword,
-      final int executeQueryMaxResultRows) {
+      final int executeQueryMaxResultRows, final int http2MultiplexingLimit) {
     this.host = Objects.requireNonNull(host);
     this.port = port;
     this.useTls = useTls;
@@ -68,6 +69,7 @@ public class ClientOptionsImpl implements ClientOptions {
     this.basicAuthUsername = basicAuthUsername;
     this.basicAuthPassword = basicAuthPassword;
     this.executeQueryMaxResultRows = executeQueryMaxResultRows;
+    this.http2MultiplexingLimit = http2MultiplexingLimit;
   }
 
   @Override
@@ -151,6 +153,12 @@ public class ClientOptionsImpl implements ClientOptions {
   }
 
   @Override
+  public ClientOptions setHttp2MultiplexingLimit(final int http2MultiplexingLimit) {
+    this.http2MultiplexingLimit = http2MultiplexingLimit;
+    return this;
+  }
+
+  @Override
   public String getHost() {
     return host == null ? "" : host;
   }
@@ -226,6 +234,9 @@ public class ClientOptionsImpl implements ClientOptions {
   }
 
   @Override
+  public int getHttp2MultiplexingLimit() { return http2MultiplexingLimit; }
+
+  @Override
   public ClientOptions copy() {
     return new ClientOptionsImpl(
         host, port,
@@ -234,7 +245,7 @@ public class ClientOptionsImpl implements ClientOptions {
         trustStorePath, trustStorePassword,
         keyStorePath, keyStorePassword, keyPassword, keyAlias,
         basicAuthUsername, basicAuthPassword,
-        executeQueryMaxResultRows);
+        executeQueryMaxResultRows, http2MultiplexingLimit);
   }
 
   // CHECKSTYLE_RULES.OFF: CyclomaticComplexity
@@ -261,14 +272,15 @@ public class ClientOptionsImpl implements ClientOptions {
         && Objects.equals(keyPassword, that.keyPassword)
         && Objects.equals(keyAlias, that.keyAlias)
         && Objects.equals(basicAuthUsername, that.basicAuthUsername)
-        && Objects.equals(basicAuthPassword, that.basicAuthPassword);
+        && Objects.equals(basicAuthPassword, that.basicAuthPassword)
+        && http2MultiplexingLimit == that.http2MultiplexingLimit;
   }
 
   @Override
   public int hashCode() {
     return Objects.hash(host, port, useTls, verifyHost, useAlpn, trustStorePath,
         trustStorePassword, keyStorePath, keyStorePassword, keyPassword, keyAlias,
-        basicAuthUsername, basicAuthPassword, executeQueryMaxResultRows);
+        basicAuthUsername, basicAuthPassword, executeQueryMaxResultRows, http2MultiplexingLimit);
   }
 
   @Override
@@ -287,7 +299,8 @@ public class ClientOptionsImpl implements ClientOptions {
         + ", keyAlias='" + keyAlias + '\''
         + ", basicAuthUsername='" + basicAuthUsername + '\''
         + ", basicAuthPassword='" + basicAuthPassword + '\''
-        + ", executeQueryMaxResultRows=" + executeQueryMaxResultRows
+        + ", executeQueryMaxResultRows=" + executeQueryMaxResultRows + '\''
+        + ", http2MultiplexingLimit=" + http2MultiplexingLimit
         + '}';
   }
 }

--- a/ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/impl/ClientOptionsImplTest.java
+++ b/ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/impl/ClientOptionsImplTest.java
@@ -63,6 +63,9 @@ public class ClientOptionsImplTest {
         .addEqualityGroup(
             ClientOptions.create().setExecuteQueryMaxResultRows(10)
         )
+        .addEqualityGroup(
+            ClientOptions.create().setHttp2MultiplexingLimit(5)
+        )
         .testEquals();
   }
 

